Add options sub spec validation

ganeshrn opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
`options` handling will support checking each individual `dict (element)` 
to resolve conditions for `aliases`, `no_log`, `mutually_exclusive`,
`required`, `type check`, `value check`, `required_together`, `required_one_of`
and `required_if` conditions in argspec. It will also set default values.

eg :
```
tasks:
  - name: Configure interface attribute with aggregate
    net_interface:
      aggregate:
        - {name: ge-0/0/1, description: test-interface-1, duplex: full, state: present}
        - {name: ge-0/0/2, description: test-interface-2, active: False}
      purge: Yes
    register: response
    
```
Usage:
```
   sub_argspec = dict(
        name=dict(),
        description=dict(no_log=True),
        duplex=dict(choices=['full', 'half', 'auto']),
        state=dict(default='present', choices=['present', 'absent', 'up', 'down']),
        active=dict(default=True, type='bool')
    )
    argument_spec = dict(
        aggregate=dict(type='list', elements='dict', options=sub_argspec),
    )
```

`purge` parameter in above example `tasks` will instruct the module to remove resources from the remote device that hasn’t been explicitly defined in aggregate. This is not supported by with_* iterators

Also, it improves performance as compared to with_* iterator for network device
that has a separate candidate and running datastore.

For with_* iteration, the sequence of operation is
```
load-config-1 (candidate db) -> commit (running db) -> load_config-2
(candidate db) -> commit (running db) ...
```
With aggregate the sequence of operation is
```
load-config-1 (candidate db) -> load-config-2 (candidate db) -> commit
(running db)
```
Since `commit` is executed  once per task for aggregate it has
huge performance benefit for large configurations.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
module_utils/basic.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.4
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

ganeshrn committed:
Add aggregate parameter validation

aggregate parameter validation will support checking each individual dict
to resolve conditions for aliases, no_log, mutually_exclusive,
required, type check, values, required_together, required_one_of
and required_if conditions in argspec. It will also set default values.

eg:
tasks:
  - name: Configure interface attribute with aggregate
    net_interface:
      aggregate:
        - {name: ge-0/0/1, description: test-interface-1, duplex: full, state: present}
        - {name: ge-0/0/2, description: test-interface-2, active: False}
    register: response
    purge: Yes

Usage:
```
from ansible.module_utils.network_common import AggregateCollection

transform = AggregateCollection(module)
param = transform(module.params.get('aggregate'))
```

Aggregate allows supports for `purge` parameter, it will instruct the module
to remove resources from remote device that hasn’t been explicitly
defined in aggregate. This is not supported by with_* iterators

Also, it improves performace as compared to with_* iterator for network device
that has seperate candidate and running datastore.
For with_* iteration the sequence of operartion is
load-config-1 (candidate db) -> commit (running db) -> load_config-2
(candidate db) -> commit (running db) ...

With aggregate the sequence of operation is
load-config-1 (candidate db) -> load-config-2 (candidate db) -> commit
(running db)

As commit is executed only once per task for aggregate it has
huge perfomance benefit for large configurations.

bcoca commented:
nice, had 1/2 of this in patch, but have not had time to finish it, quick review looks good

the issue about implementation is that we had already agreed to use a ~~subspec~~(we changed several times, i still get confused)  `options` key, which is already implemented by some modules and suported by the doc tools

ganeshrn committed:
Fix CI issues

ganeshrn commented:
Looking at the current `options` handling in `_check_argument_types()` can we move the `options` check to `__init__` and handle all `options` related things in `_handle_options()` function?
 It will handle setting aliases, set defaults, no log values, check invalid arguments, argument type, value check and optional required_*, mutually_exclusive check if `bypass_options_checks` flag is `false`.

So aggregate in the network module spec will look like:
```
    argument_spec = dict(
        < -- snip -->
        aggregate=dict(type='list', elements='dict')
        < -- snip -->
    )

argument_spec['aggregate']['options'] = argument_spec
```

bcoca commented:
yes to move options, no on having alternate implementation named 'aggregate', 'options' should be able to handle the requirements for network also

ganeshrn committed:
Fix review comments

*  Add support for options validation for aliases, no_log,
   mutually_exclusive, required, type check, value check,
   required_together, required_one_of and required_if
   conditions in sub-argspec.
*  Add unit test for options in argspec.
*  Reverted aggregate implementaion.

ganeshrn committed:
Minor change

ganeshrn commented:
bcoca: Added options validation and reverted aggregate implementation.

ganeshrn commented:
ready_for_review

bcoca commented:
cc @nitzmahone as he was also starting to develop this feature

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -773,7 +773,8 @@ class AnsibleModule(object):
                  required_one_of=None, add_file_common_args=False, supports_check_mode=False,
-                 required_if=None):
+                 required_if=None, bypass_options_checks=False, options_mutually_exclusive=None,
+                 options_required_together=None, options_required_one_of=None, options_required_if=None):
```
this seems to assume a single level of sub specs, in some cases we might want 2 or 3 so making it recursive with the top options seems like the best approach

ganeshrn replied:
For multi level sub specs if some of the args are same across different level it might result in a false error, that's why avoided having a recursive call.

ganeshrn replied:
>  making it recursive with the top options.


Does this mean `required_*` conditions or `options_required_*`?

ganeshrn replied:
bcoca: Fixed assuming module's top level options are applicable for multi-level sub args specs.

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1511,15 +1507,42 @@ def _handle_aliases(self, spec=None):
         return aliases_results
 
-    def _check_arguments(self, check_invalid_arguments):
+    def _handle_no_log_values(self, spec=None, param=None):
```
does not resolve inheritance https://github.com/ansible/ansible/issues/25097

ganeshrn replied:
Above referenced issue is not seen with this patch applied.
eg: Module arg spec
```
argument_spec = dict(
    name=dict(),
    description=dict(no_log=True),
    aggregate=dict(type='list', elements='dict'),
    purge=dict(default=False, type='bool'),
    state=dict(default='present',
               choices=['present', 'absent', 'up', 'down']),
)

aggregate_arg_spec = deepcopy(argument_spec)
del aggregate_arg_spec['aggregate']

argument_spec['aggregate']['options'] = aggregate_arg_spec
```
Module run log
```
"aggregate": [
    {
        "description": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
        "name": "ge-0/0/12",
        "purge": false,
        "state": "present",
    },
    {
        "description": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
        "name": "ge-0/0/13",
        "purge": false,
        "state": "present",
    },
    {
        "description": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
        "name": "ge-0/0/14",
        "purge": false,
        "state": "present",
    }
]
```

ganeshrn replied:
`_handle_no_log_values()` is also called from `_handle_options()` recursively in case on mullti-level argspec.
So if I understand your comment and referenced issue correctly,  this patch does resolve inheritance issue.

ganeshrn committed:
Add multi-level argspec support

*  Multi-level argspec support with module's top most
   conditionals options.

ganeshrn committed:
Fix unit test failure

ganeshrn committed:
Add parent context in errors for sub options

ganeshrn committed:
Resolve merge conflict

ganeshrn committed:
Fix CI issue

ganeshrn commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1862,6 +1916,60 @@ def _check_type_bits(self, value):
+                if spec is None or not params[k]:
+                    continue
+
+                self._options_context.append(k)
```
@nitzmahone Added ability to print parent context in case of error as discussed.

bcoca commented:
lgtm, all issues I had have been addressed, I'm probably missing something but if we merge this soon we have plenty of time to debug/fix before release.

ganeshrn commented:
bcoca: So shall I go ahead and merge this PR?

caphrim007 commented:
@ganeshrn this broke fallback functionality in the f5_utils file. Was this supposed to happen?

nitzmahone commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1862,6 +1916,60 @@ def _check_type_bits(self, value):
+
+                        self._check_required_together(self.required_together, param)
+                        self._check_required_one_of(self.required_one_of, param)
+                        self._check_required_if(self.required_if, param)
```
Upon further reflection, I think these three lines along with `check_mutually_exclusive` are erroneous for subspecs. We don't currently have a way to specify `required_if/one_of/together` or `mutually_exclusive` for subspecs, so in every case *but* the networking "aggregate" thing, this is severely broken (as you'd be making nonsensical checks in most cases)

nitzmahone replied:
Only sane way I could think of to support this properly would be to make those values reserved arg names in the arg_spec (so we could define them at any layer of the dict instead of passing them alongside). @bcoca: any other ideas?
